realm/src/main/java/io/realm/DynamicRealm.java

```diff
@@ -0,0 +1,337 @@
+
+            // Create new DynamicRealm and cache it. All exception code paths must close the DynamicRealm otherwise we risk
+            // serving faulty cache data.
+            validateAgainstExistingConfigurations(configuration);
+            boolean autoRefresh = Looper.myLooper() != null;
+            realm = new DynamicRealm(configuration, autoRefresh);
+            List<RealmConfiguration> pathConfigurationCache = globalPathConfigurationCache.get(canonicalPath);
+            if (pathConfigurationCache == null) {
+                pathConfigurationCache = new CopyOnWriteArrayList<RealmConfiguration>();
+                globalPathConfigurationCache.put(canonicalPath, pathConfigurationCache);
+            }
+            pathConfigurationCache.add(configuration);
+            realms.put(configuration, realm);
+            localRefCount.put(configuration, references + 1);
+
+            // Increment global reference counter
+            realm.acquireFileReference(configuration);
+
+            return realm;
+        }
+    }
+
+    /**
+     * Return a distinct set of objects of a specific class. As a Realm is unordered, it is undefined which objects are
+     * returned in case of multiple occurrences.
+     *
+     * @param className the Class to get objects of.
+     * @param fieldName the field name.
+     * @return A non-null {@link RealmResults} containing the distinct objects.
+     * @throws IllegalArgumentException if a field name does not exist or the field is not indexed.
+     */
+    public RealmResults<DynamicRealmObject> distinct(String className, String fieldName) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("fieldName must be provided.");
+        }
+        checkIfValid();
+        Table table = this.getTable(className);
+        long columnIndex = table.getColumnIndex(fieldName);
+        if (columnIndex == -1) {
+            throw new IllegalArgumentException(String.format("Field name '%s' does not exist.", fieldName));
+        }
+
+        TableView tableView = table.getDistinctView(columnIndex);
+        return RealmResults.createFromDynamicTableOrView(this, tableView, className);
+    }
+
+    /**
+     * Return a distinct set of objects of a specific class. As a Realm is unordered, it is undefined which objects are
+     * returned in case of multiple occurrences.
+     * This method is only available from a Looper thread.
+     *
+     * @param className the Class to get objects of.
+     * @param fieldName the field name.
+     * @return immediately an empty {@link RealmResults}. Users need to register a listener
+     * {@link io.realm.RealmResults#addChangeListener(RealmChangeListener)} to be notified
+     * when the query completes.
+     * @throws IllegalArgumentException if a field name does not exist or the field is not indexed.
+     */
+    public RealmResults<DynamicRealmObject> distinctAsync(String className, String fieldName) {
+        if (fieldName == null) {
```

Comment:
This code is repeated. Make helper. Actually I would prefer this validation code to be moved to C++ if at all possible.